keps/sig-storage/596-csi-inline-volumes/README.md

```diff
@@ -213,32 +250,214 @@ Inline volume requests can only participate in mount/unmount volume operations.
+
+  Enabling the feature gate will **not** change the default behavior.
+  Each CSI driver must opt-in to support ephemeral inline volumes.
+
+###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?
+
+  Yes. Disabling the feature gate will disable support for inline CSI volumes.
+  It can also be disabled on a specific CSI driver by removing the "Ephemeral"
+  VolumeLifecycleMode from `CSIDriver.VolumeLifecycleModes`.
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+  Reenabling the feature will again allow CSI drivers that opt-in to be used
+  as ephemeral inline volumes in the pod spec.
+
+###### Are there any tests for feature enablement/disablement?
+
+  Yes, the unit tests will test with the feature gate enabled and disabled.
+
+### Rollout, Upgrade and Rollback Planning
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+  A rollout should not impact running workloads, since the CSI driver
+  also needs to support ephemeral inline volumes, and the pod spec would need
+  to be updated to make use of the feature.
+
+  A rollback (or disabling the feature flag) may impact running workloads,
+  as any pod specs taking advantage of ephemeral inline volumes would fail
+  to start when the feature is disabled or the CSIDriver does not support it.
+  It is recommended to delete any pods using this feature before disabling it.
+
+###### What specific metrics should inform a rollback?
+
+  No rollback criteria based on metrics alone. This feature has been enabled by
+  default since 1.16 (beta) and pods that are using ephemeral inline volumes should
+  not start failing as a result of this feature moving to GA.
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+  To be documented as part of manual GA testing.
+
+<!--
+Describe manual testing that was done and the outcomes.
+Longer term, we may want to require automated upgrade/rollback tests, but we
+are missing a bunch of machinery and tooling and can't do that now.
+-->
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+  No, this feature does not deprecate any existing functionality.
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+  Check if any CSIDriver specs have `VolumeLifecycleModes` including an
+  "Ephemeral" `VolumeLifecycleMode` and if any pod specs are using this driver inline.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [x] Other (treat as last resort)
+  - Details:
+  For pods using an inline CSI driver (as in [Examples](#examples)), review
+  the status of the pod (did it mount and start successfully?).
+  Additionally, the cluster admin may review the kubelet logs for mount/unmount calls.
```

Comment:
That's not particularly scalable. If I'm managing N thousands of clusters, I can't really do that easily.

But that actually seems to me like an obvious thing for adding a metric.
Kubelet reporting a metric sth like "csi_storage_operations_errors[operation]", sounds like an obvious thing to have. And adding a label for mode (persistent/ephemeral) would allow us to actually check that easily and quickly for any cluster.

Reply:
Understood, there actually is a metric similar to what you suggest. I added it to this section and the question on SLI's. `csi_operations_seconds` allows you to filter by `driver_name`, which should be sufficient to determine if a given driver is causing a large increase in failures.